src/jobserver.h

```diff
@@ -0,0 +1,107 @@
+/// Once the command is completed, Release() must be called to return a token.
+/// The token server does not care in which order a token is received.
+struct Jobserver {
+// TODO: jobserver client support for Windows
+#ifndef _WIN32
+  /// Parse the MAKEFLAGS environment variable to receive the path / FDs / name
+  /// of the token pool, and open the handle to the pool if it is an object.
+  /// If a jobserver argument is found in the MAKEFLAGS environment variable,
+  /// and the handle is successfully opened, later calls to Enable() return true.
+  /// If a jobserver argument is found, but the handle fails to be opened,
+  /// the ninja process is aborted with an error.
+  Jobserver();
+
+  /// Before exiting Ninja, ensure that tokens are returned and handles closed.
+  ~Jobserver();
+
+  /// Return current token count or initialization signal if negative.
+  int Tokens() const { return token_count_; }
+
+  /// Return true if jobserver functionality is enabled and initialized.
+  bool Enabled() const { return rfd_ >= 0 && wfd_ >= 0; }
+#endif
+
+  /// Implementation-specific method to acquire a token from the external pool
+  /// which is called for all tokens but returns early for the first token.
+  /// This method is called every time Ninja needs to start a command process.
+  /// Return true on success (token acquired), and false on failure (no tokens
+  /// available). First call always succeeds. Ninja is aborted on read errors.
+  bool Acquire();
+
+  /// Implementation-specific method to release a token to the external pool
+  /// which is called for all tokens but returns early for the last token.
+  /// Return a previously acquired token to the external token pool.
+  /// It must be called for each successful call to Acquire() after the command
+  /// even if subprocesses fail or in the case of errors causing Ninja to exit.
+  /// Ninja is aborted on write errors, and otherwise calls always succeed.
+  void Release();
+
+  /// Loop through Release() to return all tokens. Called before Ninja exits.
+  void Clear();
+
+ private:
+  /// The number of currently acquired tokens, or a status signal if negative.
+  /// Used to verify that all acquired tokens have been released before exiting,
+  /// and when the implicit (first) token has been acquired (initialization).
+  /// -1: initialized without a token
+  ///  0: uninitialized or disabled
+  /// +n: number of tokens in use
+  int token_count_ = 0;
+
+  /// String of the parsed value of the jobserver flag passed to environment.
+  std::string jobserver_name_;
+
+// TODO: jobserver client support for Windows
+#ifdef _WIN32
+#else
+  /// Whether the type of jobserver pipe supplied to ninja is named.
+  bool jobserver_fifo_ = false;
+
+  /// File descriptors to communicate with upstream jobserver token pool.
+  int rfd_ = -1;
+  int wfd_ = -1;
+#endif
+
+  /// Helpers for initialization.
+  std::string::size_type flag_char_ = 0;
+  std::string flag_;
+  std::vector<std::string> flags_;
```

Comment:
Move these to be local variables in the constructor instead of member variables.

Reply:
I can try, but earlier attempts to do something similar before I came up with this resulted in segfault when using the value of a local variable to define a member variable...

Reply:
segfault would always happen in all cases if the variables are not declared/instantiated correctly, and would never happen if they are.

Reply:
I split out the text parsing to its own function, and added a unit test in https://github.com/ninja-build/ninja/commit/8580d5142eb2b01b9e327d52d86d8de454713472
What do you think? I can clean up the code if you agree that this is the way to go.

Reply:
> I split out the text parsing to its own function, and added a unit test in [8580d51]

its a nice start for unit tests, but I don't think we want to introduce tuples to ninja when it's not necessary. ***maybe*** a `std::pair` for the FDs but I still don't see it as necessary.